### PR TITLE
drivers: serial: ns16550: fill full fifo

### PR DESCRIPTION
**Pull request by niedzwiecki-dawid**
Put the maximum number of bytes into Tx FIFO in the fill_fifo routine
to reduce CPU usage.

Previously, the THRE bit was checked in a loop, but, according to doc -
"In the FIFO mode, it is set when the XMIT FIFO is empty, and is
cleared when at least one byte is written to XMIT FIFO.", so only one
byte was transferred every interrupt. That was generating a big amount
of interrupts, which consumes CPU time.

Signed-off-by: Dawid Niedzwiecki <dn@semihalf.com>

### CODE REVIEW

File: drivers/serial/uart_ns16550.c [hunk 1]
```diff
@@ -601,7 +612,7 @@ static int uart_ns16550_fifo_fill(const struct device *dev,
 	int i;
 	k_spinlock_key_t key = k_spin_lock(&DEV_DATA(dev)->lock);
 
-	for (i = 0; (i < size) && (INBYTE(LSR(dev)) & LSR_THRE) != 0; i++) {
+	for (i = 0; (i < size) && i < DEV_DATA(dev)->fifo_size; i++) {
```
Comment (keith-zephyr) [inline drivers/serial/uart_ns16550.c:615]: Should this check the LSR_THRE bit once before the loop?

Comment (semihalf-niedzwiecki-dawid) [reply drivers/serial/uart_ns16550.c:615]: The fifo_fill function can be called only from ISR after uart_irq_tx_ready() returning true. AFAIK and according to tests, the interrupt is triggered when the FIFO is empty, so IMO checking it one more time is not necessary.

Comment (keith-zephyr) [reply drivers/serial/uart_ns16550.c:615]: Sounds good. The ``i < fifo_size`` check should probably have enclosing parenthesis to match the existing code style.

File: drivers/serial/uart_ns16550.c [hunk 2]
```diff
@@ -464,6 +466,15 @@ static int uart_ns16550_configure(const struct device *dev,
...
 
+	if ((INBYTE(IIR(dev)) & IIR_FE) == IIR_FE)
+#ifdef CONFIG_UART_NS16750
+		dev_data->fifo_size = 64;
+#else
+		dev_data->fifo_size = 16;
+#endif
+	else
+		dev_data->fifo_size = 1;
+
```
Comment (dcpleung) [inline drivers/serial/uart_ns16550.c:477]: Brackets are needed even for a one line statement.

**Review by dcpleung — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_ns16550.c (hunk 1 above)
Comment (semihalf-niedzwiecki-dawid) [reply drivers/serial/uart_ns16550.c]: Done.

File: drivers/serial/uart_ns16550.c (hunk 2 above)
Comment (semihalf-niedzwiecki-dawid) [reply drivers/serial/uart_ns16550.c]: Done.

**Review by keith-zephyr — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Pull request merged**